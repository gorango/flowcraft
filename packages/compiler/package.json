{
	"name": "@flowcraft/compiler",
	"type": "module",
	"version": "1.0.0",
	"description": "A compiler for transforming imperative TypeScript into declarative Flowcraft workflows",
	"license": "MIT",
	"homepage": "https://flowcraft.js.org",
	"repository": {
		"type": "git",
		"url": "https://github.com/gorango/flowcraft.git"
	},
	"bugs": "https://github.com/gorango/flowcraft/issues",
	"keywords": [
		"flowcraft",
		"compiler",
		"workflow",
		"typescript"
	],
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"import": "./dist/index.js"
		}
	},
	"files": [
		"dist"
	],
	"scripts": {
		"dev": "tsup --watch",
		"build": "tsup",
		"typecheck": "tsc --noEmit",
		"test": "vitest run",
		"test:watch": "vitest",
		"test:coverage": "vitest run --coverage"
	},
	"dependencies": {
		"flowcraft": "workspace:*",
		"resolve": "^1.22.11",
		"typescript": "^5.9.3"
	},
	"devDependencies": {
		"@types/node": "^24.9.1",
		"@types/resolve": "^1.20.6",
		"@vitest/coverage-v8": "3.2.4",
		"tsup": "^8.5.0",
		"vitest": "^4.0.4"
	}
}
