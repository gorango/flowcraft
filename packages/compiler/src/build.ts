import fs from 'node:fs/promises'
import path from 'node:path'
import { compileProject } from './index'

export interface CompileFlowsOptions {
	entryPoints?: string[]
	tsConfigPath?: string
	manifestPath?: string
}

// This is the generic, reusable function
export async function buildFlows(options: CompileFlowsOptions = {}) {
	const projectRoot = process.cwd()
	const entryPoints = options.entryPoints ?? [path.resolve(projectRoot, 'src/index.ts')]
	const tsConfigPath = options.tsConfigPath ?? path.resolve(projectRoot, 'tsconfig.json')
	const manifestPath = options.manifestPath ?? path.resolve(projectRoot, 'dist/flowcraft.manifest.js')

	console.log('Compiling Flowcraft workflows...')

	try {
		const { diagnostics, manifestSource } = compileProject(entryPoints, tsConfigPath)

		if (diagnostics.some((d) => d.severity === 'error')) {
			console.error('❌ Flowcraft compilation failed:')
			diagnostics
				.filter((d) => d.severity === 'error')
				.forEach((d) => {
					console.error(`  - ${path.relative(projectRoot, d.file)}:${d.line}:${d.column} - ${d.message}`)
				})
			throw new Error('Flowcraft compilation failed.')
		}

		await fs.mkdir(path.dirname(manifestPath), { recursive: true })
		await fs.writeFile(manifestPath, manifestSource)
		console.log(`✅ Flowcraft compilation successful! Manifest: ${path.relative(projectRoot, manifestPath)}`)
	} catch (error) {
		console.error('❌ An unexpected error occurred during Flowcraft compilation:')
		// Don't log the whole error object unless verbose logging is on
		console.error(error instanceof Error ? error.message : String(error))
		// Re-throw to fail the build process
		throw error
	}
}
