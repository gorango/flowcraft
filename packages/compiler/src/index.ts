import { Compiler } from './compiler'
import type { CompilationOutput } from './types'

export function compileProject(entryFilePaths: string[], tsConfigPath: string): CompilationOutput {
	const compiler = new Compiler(tsConfigPath)
	return compiler.compileProject(entryFilePaths)
}
