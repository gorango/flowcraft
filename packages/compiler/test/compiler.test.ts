import { describe, expect, it } from 'vitest'
import { Compiler } from '../src/compiler'

describe('Compiler', () => {
	it('should compile a simple project', () => {
		const compiler = new Compiler('tsconfig.json')
		const result = compiler.compileProject(['src/index.ts'])
		expect(result.blueprints).toBeDefined()
		expect(result.registry).toBeDefined()
		expect(result.diagnostics).toBeDefined()
	})

	it('should compile all test fixtures', () => {
		const compiler = new Compiler('tsconfig.json')
		const result = compiler.compileProject(['test/fixtures/index.ts'])
		expect(result.blueprints).toMatchSnapshot()
		expect(result.registry).toMatchSnapshot()
		expect(result.manifestSource).toMatchSnapshot()
		expect(result.diagnostics).toMatchSnapshot()
	})
})
