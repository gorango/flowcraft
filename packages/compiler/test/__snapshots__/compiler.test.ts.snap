// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Compiler > should compile all test fixtures 1`] = `
{
  "complexControlFlow": {
    "edges": [
      {
        "condition": "!validated.isValid",
        "source": "validateData_1",
        "target": "handleValidationError_1",
      },
      {
        "source": "validateData_1",
        "target": "loop-controller_1",
      },
      {
        "source": "handleValidationError_1",
        "target": "handleBatchError_1",
      },
      {
        "condition": "!(!validated.isValid)",
        "source": "validateData_1",
        "target": "handleBatchError_1",
      },
      {
        "action": "break",
        "source": "loop-controller_1",
        "target": "handleBatchError_1",
      },
      {
        "source": "processBatch_1",
        "target": "handleBatchError_1",
      },
      {
        "action": "continue",
        "source": "loop-controller_1",
        "target": "processBatch_1",
      },
      {
        "source": "handleBatchError_1",
        "target": "loop-controller_1",
      },
      {
        "source": "loop-controller_1",
        "target": "generateSummary_parallel_1",
      },
      {
        "source": "loop-controller_1",
        "target": "generateReport_parallel_1",
      },
      {
        "source": "generateSummary_parallel_1",
        "target": "gather_1",
      },
      {
        "source": "generateReport_parallel_1",
        "target": "gather_1",
      },
      {
        "source": "processBatch_1",
        "target": "sendHighVolumeNotification_1",
      },
      {
        "source": "handleBatchError_1",
        "target": "sendHighVolumeNotification_1",
      },
      {
        "condition": "processedCount > 1000",
        "source": "gather_1",
        "target": "sendHighVolumeNotification_1",
      },
      {
        "condition": "!(processedCount > 1000)",
        "source": "gather_1",
        "target": "sendStandardNotification_1",
      },
    ],
    "id": "complexControlFlow",
    "nodes": [
      {
        "id": "validateData_1",
        "uses": "validateData",
      },
      {
        "id": "handleValidationError_1",
        "uses": "handleValidationError",
      },
      {
        "config": {
          "joinStrategy": "any",
        },
        "id": "loop-controller_1",
        "params": {
          "condition": "currentBatch && currentBatch.length > 0",
        },
        "uses": "loop-controller",
      },
      {
        "config": {
          "fallback": "handleBatchError_1",
        },
        "id": "processBatch_1",
        "uses": "processBatch",
      },
      {
        "id": "handleBatchError_1",
        "uses": "handleBatchError",
      },
      {
        "id": "generateSummary_parallel_1",
        "uses": "generateSummary",
      },
      {
        "id": "generateReport_parallel_1",
        "uses": "generateReport",
      },
      {
        "config": {
          "joinStrategy": "all",
        },
        "id": "gather_1",
        "uses": "gather",
      },
      {
        "config": {
          "joinStrategy": "any",
        },
        "id": "sendHighVolumeNotification_1",
        "uses": "sendHighVolumeNotification",
      },
      {
        "id": "sendStandardNotification_1",
        "uses": "sendStandardNotification",
      },
    ],
  },
  "complexParallelFlow": {
    "edges": [
      {
        "source": "processItems_parallel_1",
        "target": "gather_1",
      },
      {
        "source": "validateItems_parallel_1",
        "target": "gather_1",
      },
      {
        "source": "enrichItems_parallel_1",
        "target": "gather_1",
      },
      {
        "condition": "validated.valid",
        "source": "gather_1",
        "target": "saveValidItems_1",
      },
      {
        "condition": "!(validated.valid)",
        "source": "gather_1",
        "target": "handleInvalidItems_1",
      },
      {
        "source": "saveValidItems_1",
        "target": "finalizeResults_1",
      },
      {
        "source": "handleInvalidItems_1",
        "target": "finalizeResults_1",
      },
      {
        "source": "gather_1",
        "target": "finalizeResults_1",
      },
    ],
    "id": "complexParallelFlow",
    "nodes": [
      {
        "id": "processItems_parallel_1",
        "uses": "processItems",
      },
      {
        "id": "validateItems_parallel_1",
        "uses": "validateItems",
      },
      {
        "id": "enrichItems_parallel_1",
        "uses": "enrichItems",
      },
      {
        "config": {
          "joinStrategy": "all",
        },
        "id": "gather_1",
        "uses": "gather",
      },
      {
        "id": "saveValidItems_1",
        "uses": "saveValidItems",
      },
      {
        "id": "handleInvalidItems_1",
        "uses": "handleInvalidItems",
      },
      {
        "config": {
          "joinStrategy": "any",
        },
        "id": "finalizeResults_1",
        "uses": "finalizeResults",
      },
    ],
  },
  "mainFlow": {
    "edges": [
      {
        "source": "fetchUser_1",
        "target": "subFlow_1",
      },
      {
        "source": "subFlow_1",
        "target": "processOrders_1",
      },
    ],
    "id": "mainFlow",
    "nodes": [
      {
        "id": "fetchUser_1",
        "uses": "fetchUser",
      },
      {
        "id": "subFlow_1",
        "params": {
          "blueprintId": "subFlow",
        },
        "uses": "subflow",
      },
      {
        "id": "processOrders_1",
        "uses": "processOrders",
      },
    ],
  },
  "nestedControlFlow": {
    "edges": [
      {
        "condition": "reviewed.approved",
        "source": "reviewItem_1",
        "target": "approveItem_1",
      },
      {
        "condition": "!(reviewed.approved)",
        "source": "reviewItem_1",
        "target": "rejectItem_1",
      },
      {
        "condition": "processed.needsReview",
        "source": "processItem_1",
        "target": "reviewItem_1",
      },
      {
        "source": "approveItem_1",
        "target": "autoApproveItem_1",
      },
      {
        "source": "rejectItem_1",
        "target": "autoApproveItem_1",
      },
      {
        "source": "autoApproveItem_1",
        "target": "flagItemForManualReview_1",
      },
      {
        "condition": "!(processed.needsReview)",
        "source": "processItem_1",
        "target": "autoApproveItem_1",
      },
      {
        "source": "reviewItem_1",
        "target": "finalizeProcessing_1",
      },
      {
        "source": "flagItemForManualReview_1",
        "target": "finalizeProcessing_1",
      },
      {
        "source": "processItem_1",
        "target": "finalizeProcessing_1",
      },
    ],
    "id": "nestedControlFlow",
    "nodes": [
      {
        "id": "processItem_1",
        "uses": "processItem",
      },
      {
        "id": "reviewItem_1",
        "uses": "reviewItem",
      },
      {
        "id": "approveItem_1",
        "uses": "approveItem",
      },
      {
        "id": "rejectItem_1",
        "uses": "rejectItem",
      },
      {
        "config": {
          "joinStrategy": "any",
        },
        "id": "autoApproveItem_1",
        "uses": "autoApproveItem",
      },
      {
        "id": "flagItemForManualReview_1",
        "uses": "flagItemForManualReview",
      },
      {
        "config": {
          "joinStrategy": "any",
        },
        "id": "finalizeProcessing_1",
        "uses": "finalizeProcessing",
      },
    ],
  },
  "parallelFlow": {
    "edges": [
      {
        "source": "fetchProfile_parallel_1",
        "target": "aggregateData_parallel_1",
      },
      {
        "source": "fetchOrders_parallel_1",
        "target": "aggregateData_parallel_1",
      },
      {
        "source": "fetchActivity_parallel_1",
        "target": "aggregateData_parallel_1",
      },
      {
        "source": "aggregateData_parallel_1",
        "target": "aggregateData_2",
      },
    ],
    "id": "parallelFlow",
    "nodes": [
      {
        "id": "fetchProfile_parallel_1",
        "uses": "fetchProfile",
      },
      {
        "id": "fetchOrders_parallel_1",
        "uses": "fetchOrders",
      },
      {
        "id": "fetchActivity_parallel_1",
        "uses": "fetchActivity",
      },
      {
        "config": {
          "joinStrategy": "all",
        },
        "id": "aggregateData_parallel_1",
        "uses": "aggregateData",
      },
      {
        "id": "aggregateData_2",
        "uses": "aggregateData",
      },
    ],
  },
  "simpleFlow": {
    "edges": [
      {
        "source": "fetchUser_1",
        "target": "processOrders_1",
      },
    ],
    "id": "simpleFlow",
    "nodes": [
      {
        "id": "fetchUser_1",
        "uses": "fetchUser",
      },
      {
        "id": "processOrders_1",
        "uses": "processOrders",
      },
    ],
  },
  "simpleIfElseFlow": {
    "edges": [
      {
        "condition": "condition",
        "source": "start",
        "target": "doSomething_1",
      },
      {
        "condition": "!(condition)",
        "source": "start",
        "target": "doSomethingElse_1",
      },
      {
        "source": "doSomething_1",
        "target": "finalize_1",
      },
      {
        "source": "doSomethingElse_1",
        "target": "finalize_1",
      },
    ],
    "id": "simpleIfElseFlow",
    "nodes": [
      {
        "id": "start",
        "uses": "start",
      },
      {
        "id": "doSomething_1",
        "uses": "doSomething",
      },
      {
        "id": "doSomethingElse_1",
        "uses": "doSomethingElse",
      },
      {
        "config": {
          "joinStrategy": "any",
        },
        "id": "finalize_1",
        "uses": "finalize",
      },
    ],
  },
  "simpleParallelFlow": {
    "edges": [
      {
        "source": "fetchProfile_parallel_1",
        "target": "aggregateData_parallel_1",
      },
      {
        "source": "fetchOrders_parallel_1",
        "target": "aggregateData_parallel_1",
      },
      {
        "source": "aggregateData_parallel_1",
        "target": "aggregateData_2",
      },
    ],
    "id": "simpleParallelFlow",
    "nodes": [
      {
        "id": "fetchProfile_parallel_1",
        "uses": "fetchProfile",
      },
      {
        "id": "fetchOrders_parallel_1",
        "uses": "fetchOrders",
      },
      {
        "config": {
          "joinStrategy": "all",
        },
        "id": "aggregateData_parallel_1",
        "uses": "aggregateData",
      },
      {
        "id": "aggregateData_2",
        "uses": "aggregateData",
      },
    ],
  },
  "subFlow": {
    "edges": [],
    "id": "subFlow",
    "nodes": [
      {
        "id": "recordTransaction_1",
        "uses": "recordTransaction",
      },
    ],
  },
}
`;

exports[`Compiler > should compile all test fixtures 2`] = `
{
  "aggregateData": {
    "exportName": "aggregateData",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-parallel.ts",
  },
  "approveItem": {
    "exportName": "approveItem",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "autoApproveItem": {
    "exportName": "autoApproveItem",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "doSomething": {
    "exportName": "doSomething",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-if-else.ts",
  },
  "doSomethingElse": {
    "exportName": "doSomethingElse",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-if-else.ts",
  },
  "enrichItems": {
    "exportName": "enrichItems",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/parallel-flow.ts",
  },
  "fetchActivity": {
    "exportName": "fetchActivity",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/parallel-flow.ts",
  },
  "fetchOrders": {
    "exportName": "fetchOrders",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-parallel.ts",
  },
  "fetchProfile": {
    "exportName": "fetchProfile",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-parallel.ts",
  },
  "fetchUser": {
    "exportName": "fetchUser",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-flow.ts",
  },
  "finalize": {
    "exportName": "finalize",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-if-else.ts",
  },
  "finalizeProcessing": {
    "exportName": "finalizeProcessing",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "finalizeResults": {
    "exportName": "finalizeResults",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/parallel-flow.ts",
  },
  "flagItemForManualReview": {
    "exportName": "flagItemForManualReview",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "generateReport": {
    "exportName": "generateReport",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "generateSummary": {
    "exportName": "generateSummary",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "handleBatchError": {
    "exportName": "handleBatchError",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "handleInvalidItems": {
    "exportName": "handleInvalidItems",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/parallel-flow.ts",
  },
  "handleValidationError": {
    "exportName": "handleValidationError",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "processBatch": {
    "exportName": "processBatch",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "processItem": {
    "exportName": "processItem",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "processItems": {
    "exportName": "processItems",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/parallel-flow.ts",
  },
  "processOrders": {
    "exportName": "processOrders",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/simple-flow.ts",
  },
  "recordTransaction": {
    "exportName": "recordTransaction",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/steps.ts",
  },
  "rejectItem": {
    "exportName": "rejectItem",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "reviewItem": {
    "exportName": "reviewItem",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "saveValidItems": {
    "exportName": "saveValidItems",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/parallel-flow.ts",
  },
  "sendHighVolumeNotification": {
    "exportName": "sendHighVolumeNotification",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "sendStandardNotification": {
    "exportName": "sendStandardNotification",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "validateData": {
    "exportName": "validateData",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/complex-control-flow.ts",
  },
  "validateItems": {
    "exportName": "validateItems",
    "importPath": "/home/g/m/flowcraft/packages/compiler/test/fixtures/parallel-flow.ts",
  },
}
`;

exports[`Compiler > should compile all test fixtures 3`] = `
"// Generated by @flowcraft/compiler
import { validateData } from '../test/fixtures/complex-control-flow'
import { handleValidationError } from '../test/fixtures/complex-control-flow'
import { handleBatchError } from '../test/fixtures/complex-control-flow'
import { processBatch } from '../test/fixtures/complex-control-flow'
import { generateSummary } from '../test/fixtures/complex-control-flow'
import { generateReport } from '../test/fixtures/complex-control-flow'
import { sendHighVolumeNotification } from '../test/fixtures/complex-control-flow'
import { sendStandardNotification } from '../test/fixtures/complex-control-flow'
import { processItem } from '../test/fixtures/complex-control-flow'
import { reviewItem } from '../test/fixtures/complex-control-flow'
import { approveItem } from '../test/fixtures/complex-control-flow'
import { rejectItem } from '../test/fixtures/complex-control-flow'
import { autoApproveItem } from '../test/fixtures/complex-control-flow'
import { flagItemForManualReview } from '../test/fixtures/complex-control-flow'
import { finalizeProcessing } from '../test/fixtures/complex-control-flow'
import { recordTransaction } from '../test/fixtures/steps'
import { fetchUser } from '../test/fixtures/simple-flow'
import { processOrders } from '../test/fixtures/simple-flow'
import { fetchProfile } from '../test/fixtures/simple-parallel'
import { fetchOrders } from '../test/fixtures/simple-parallel'
import { fetchActivity } from '../test/fixtures/parallel-flow'
import { aggregateData } from '../test/fixtures/simple-parallel'
import { processItems } from '../test/fixtures/parallel-flow'
import { validateItems } from '../test/fixtures/parallel-flow'
import { enrichItems } from '../test/fixtures/parallel-flow'
import { saveValidItems } from '../test/fixtures/parallel-flow'
import { handleInvalidItems } from '../test/fixtures/parallel-flow'
import { finalizeResults } from '../test/fixtures/parallel-flow'
import { doSomething } from '../test/fixtures/simple-if-else'
import { doSomethingElse } from '../test/fixtures/simple-if-else'
import { finalize } from '../test/fixtures/simple-if-else'

import type { NodeImplementation, WorkflowBlueprint } from 'flowcraft'

export const registry: Record<string, NodeImplementation> = {
  'validateData': validateData,
  'handleValidationError': handleValidationError,
  'handleBatchError': handleBatchError,
  'processBatch': processBatch,
  'generateSummary': generateSummary,
  'generateReport': generateReport,
  'sendHighVolumeNotification': sendHighVolumeNotification,
  'sendStandardNotification': sendStandardNotification,
  'processItem': processItem,
  'reviewItem': reviewItem,
  'approveItem': approveItem,
  'rejectItem': rejectItem,
  'autoApproveItem': autoApproveItem,
  'flagItemForManualReview': flagItemForManualReview,
  'finalizeProcessing': finalizeProcessing,
  'recordTransaction': recordTransaction,
  'fetchUser': fetchUser,
  'processOrders': processOrders,
  'fetchProfile': fetchProfile,
  'fetchOrders': fetchOrders,
  'fetchActivity': fetchActivity,
  'aggregateData': aggregateData,
  'processItems': processItems,
  'validateItems': validateItems,
  'enrichItems': enrichItems,
  'saveValidItems': saveValidItems,
  'handleInvalidItems': handleInvalidItems,
  'finalizeResults': finalizeResults,
  'doSomething': doSomething,
  'doSomethingElse': doSomethingElse,
  'finalize': finalize
}

export const blueprints: Record<string, WorkflowBlueprint> = {
  'complexControlFlow': {
  "id": "complexControlFlow",
  "nodes": [
    {
      "id": "validateData_1",
      "uses": "validateData"
    },
    {
      "id": "handleValidationError_1",
      "uses": "handleValidationError"
    },
    {
      "id": "loop-controller_1",
      "uses": "loop-controller",
      "params": {
        "condition": "currentBatch && currentBatch.length > 0"
      },
      "config": {
        "joinStrategy": "any"
      }
    },
    {
      "id": "processBatch_1",
      "uses": "processBatch",
      "config": {
        "fallback": "handleBatchError_1"
      }
    },
    {
      "id": "handleBatchError_1",
      "uses": "handleBatchError"
    },
    {
      "id": "generateSummary_parallel_1",
      "uses": "generateSummary"
    },
    {
      "id": "generateReport_parallel_1",
      "uses": "generateReport"
    },
    {
      "id": "gather_1",
      "uses": "gather",
      "config": {
        "joinStrategy": "all"
      }
    },
    {
      "id": "sendHighVolumeNotification_1",
      "uses": "sendHighVolumeNotification",
      "config": {
        "joinStrategy": "any"
      }
    },
    {
      "id": "sendStandardNotification_1",
      "uses": "sendStandardNotification"
    }
  ],
  "edges": [
    {
      "source": "validateData_1",
      "target": "handleValidationError_1",
      "condition": "!validated.isValid"
    },
    {
      "source": "validateData_1",
      "target": "loop-controller_1"
    },
    {
      "source": "handleValidationError_1",
      "target": "handleBatchError_1"
    },
    {
      "source": "validateData_1",
      "target": "handleBatchError_1",
      "condition": "!(!validated.isValid)"
    },
    {
      "source": "loop-controller_1",
      "target": "handleBatchError_1",
      "action": "break"
    },
    {
      "source": "processBatch_1",
      "target": "handleBatchError_1"
    },
    {
      "source": "loop-controller_1",
      "target": "processBatch_1",
      "action": "continue"
    },
    {
      "source": "handleBatchError_1",
      "target": "loop-controller_1"
    },
    {
      "source": "loop-controller_1",
      "target": "generateSummary_parallel_1"
    },
    {
      "source": "loop-controller_1",
      "target": "generateReport_parallel_1"
    },
    {
      "source": "generateSummary_parallel_1",
      "target": "gather_1"
    },
    {
      "source": "generateReport_parallel_1",
      "target": "gather_1"
    },
    {
      "source": "processBatch_1",
      "target": "sendHighVolumeNotification_1"
    },
    {
      "source": "handleBatchError_1",
      "target": "sendHighVolumeNotification_1"
    },
    {
      "source": "gather_1",
      "target": "sendHighVolumeNotification_1",
      "condition": "processedCount > 1000"
    },
    {
      "source": "gather_1",
      "target": "sendStandardNotification_1",
      "condition": "!(processedCount > 1000)"
    }
  ]
},
  'nestedControlFlow': {
  "id": "nestedControlFlow",
  "nodes": [
    {
      "id": "processItem_1",
      "uses": "processItem"
    },
    {
      "id": "reviewItem_1",
      "uses": "reviewItem"
    },
    {
      "id": "approveItem_1",
      "uses": "approveItem"
    },
    {
      "id": "rejectItem_1",
      "uses": "rejectItem"
    },
    {
      "id": "autoApproveItem_1",
      "uses": "autoApproveItem",
      "config": {
        "joinStrategy": "any"
      }
    },
    {
      "id": "flagItemForManualReview_1",
      "uses": "flagItemForManualReview"
    },
    {
      "id": "finalizeProcessing_1",
      "uses": "finalizeProcessing",
      "config": {
        "joinStrategy": "any"
      }
    }
  ],
  "edges": [
    {
      "source": "reviewItem_1",
      "target": "approveItem_1",
      "condition": "reviewed.approved"
    },
    {
      "source": "reviewItem_1",
      "target": "rejectItem_1",
      "condition": "!(reviewed.approved)"
    },
    {
      "source": "processItem_1",
      "target": "reviewItem_1",
      "condition": "processed.needsReview"
    },
    {
      "source": "approveItem_1",
      "target": "autoApproveItem_1"
    },
    {
      "source": "rejectItem_1",
      "target": "autoApproveItem_1"
    },
    {
      "source": "autoApproveItem_1",
      "target": "flagItemForManualReview_1"
    },
    {
      "source": "processItem_1",
      "target": "autoApproveItem_1",
      "condition": "!(processed.needsReview)"
    },
    {
      "source": "reviewItem_1",
      "target": "finalizeProcessing_1"
    },
    {
      "source": "flagItemForManualReview_1",
      "target": "finalizeProcessing_1"
    },
    {
      "source": "processItem_1",
      "target": "finalizeProcessing_1"
    }
  ]
},
  'subFlow': {
  "id": "subFlow",
  "nodes": [
    {
      "id": "recordTransaction_1",
      "uses": "recordTransaction"
    }
  ],
  "edges": []
},
  'mainFlow': {
  "id": "mainFlow",
  "nodes": [
    {
      "id": "fetchUser_1",
      "uses": "fetchUser"
    },
    {
      "id": "subFlow_1",
      "uses": "subflow",
      "params": {
        "blueprintId": "subFlow"
      }
    },
    {
      "id": "processOrders_1",
      "uses": "processOrders"
    }
  ],
  "edges": [
    {
      "source": "fetchUser_1",
      "target": "subFlow_1"
    },
    {
      "source": "subFlow_1",
      "target": "processOrders_1"
    }
  ]
},
  'parallelFlow': {
  "id": "parallelFlow",
  "nodes": [
    {
      "id": "fetchProfile_parallel_1",
      "uses": "fetchProfile"
    },
    {
      "id": "fetchOrders_parallel_1",
      "uses": "fetchOrders"
    },
    {
      "id": "fetchActivity_parallel_1",
      "uses": "fetchActivity"
    },
    {
      "id": "aggregateData_parallel_1",
      "uses": "aggregateData",
      "config": {
        "joinStrategy": "all"
      }
    },
    {
      "id": "aggregateData_2",
      "uses": "aggregateData"
    }
  ],
  "edges": [
    {
      "source": "fetchProfile_parallel_1",
      "target": "aggregateData_parallel_1"
    },
    {
      "source": "fetchOrders_parallel_1",
      "target": "aggregateData_parallel_1"
    },
    {
      "source": "fetchActivity_parallel_1",
      "target": "aggregateData_parallel_1"
    },
    {
      "source": "aggregateData_parallel_1",
      "target": "aggregateData_2"
    }
  ]
},
  'complexParallelFlow': {
  "id": "complexParallelFlow",
  "nodes": [
    {
      "id": "processItems_parallel_1",
      "uses": "processItems"
    },
    {
      "id": "validateItems_parallel_1",
      "uses": "validateItems"
    },
    {
      "id": "enrichItems_parallel_1",
      "uses": "enrichItems"
    },
    {
      "id": "gather_1",
      "uses": "gather",
      "config": {
        "joinStrategy": "all"
      }
    },
    {
      "id": "saveValidItems_1",
      "uses": "saveValidItems"
    },
    {
      "id": "handleInvalidItems_1",
      "uses": "handleInvalidItems"
    },
    {
      "id": "finalizeResults_1",
      "uses": "finalizeResults",
      "config": {
        "joinStrategy": "any"
      }
    }
  ],
  "edges": [
    {
      "source": "processItems_parallel_1",
      "target": "gather_1"
    },
    {
      "source": "validateItems_parallel_1",
      "target": "gather_1"
    },
    {
      "source": "enrichItems_parallel_1",
      "target": "gather_1"
    },
    {
      "source": "gather_1",
      "target": "saveValidItems_1",
      "condition": "validated.valid"
    },
    {
      "source": "gather_1",
      "target": "handleInvalidItems_1",
      "condition": "!(validated.valid)"
    },
    {
      "source": "saveValidItems_1",
      "target": "finalizeResults_1"
    },
    {
      "source": "handleInvalidItems_1",
      "target": "finalizeResults_1"
    },
    {
      "source": "gather_1",
      "target": "finalizeResults_1"
    }
  ]
},
  'simpleFlow': {
  "id": "simpleFlow",
  "nodes": [
    {
      "id": "fetchUser_1",
      "uses": "fetchUser"
    },
    {
      "id": "processOrders_1",
      "uses": "processOrders"
    }
  ],
  "edges": [
    {
      "source": "fetchUser_1",
      "target": "processOrders_1"
    }
  ]
},
  'simpleIfElseFlow': {
  "id": "simpleIfElseFlow",
  "nodes": [
    {
      "id": "start",
      "uses": "start"
    },
    {
      "id": "doSomething_1",
      "uses": "doSomething"
    },
    {
      "id": "doSomethingElse_1",
      "uses": "doSomethingElse"
    },
    {
      "id": "finalize_1",
      "uses": "finalize",
      "config": {
        "joinStrategy": "any"
      }
    }
  ],
  "edges": [
    {
      "source": "start",
      "target": "doSomething_1",
      "condition": "condition"
    },
    {
      "source": "start",
      "target": "doSomethingElse_1",
      "condition": "!(condition)"
    },
    {
      "source": "doSomething_1",
      "target": "finalize_1"
    },
    {
      "source": "doSomethingElse_1",
      "target": "finalize_1"
    }
  ]
},
  'simpleParallelFlow': {
  "id": "simpleParallelFlow",
  "nodes": [
    {
      "id": "fetchProfile_parallel_1",
      "uses": "fetchProfile"
    },
    {
      "id": "fetchOrders_parallel_1",
      "uses": "fetchOrders"
    },
    {
      "id": "aggregateData_parallel_1",
      "uses": "aggregateData",
      "config": {
        "joinStrategy": "all"
      }
    },
    {
      "id": "aggregateData_2",
      "uses": "aggregateData"
    }
  ],
  "edges": [
    {
      "source": "fetchProfile_parallel_1",
      "target": "aggregateData_parallel_1"
    },
    {
      "source": "fetchOrders_parallel_1",
      "target": "aggregateData_parallel_1"
    },
    {
      "source": "aggregateData_parallel_1",
      "target": "aggregateData_2"
    }
  ]
}
}
"
`;
